### ISSUE TITLE
WASM/WebGL2

### ISSUE THREAD

**Issue by bonsairobo** [bonsairobo/bevy_triplanar_splatting#2]
Currently trying to run this for target `wasm32-unknown-unknown` will fail to render anything, even if proper features and texture formats (KTX2) are used.

Browser log output:

```
INFO /home/duncan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bevy_winit-0.12.0/src/system.rs:55 Creating new window "App" (0v0) [wasm.js:344:9](http://127.0.0.1:1334/api/wasm.js)
WEBGL_debug_renderer_info is deprecated in Firefox and will be removed. Please use RENDERER. [wasm.js:738:18](http://127.0.0.1:1334/api/wasm.js)
INFO /home/duncan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bevy_render-0.12.0/src/renderer/mod.rs:140 AdapterInfo { name: "Radeon HD 3200 Graphics", vendor: 4098, device: 0, device_type: IntegratedGpu, driver: "", driver_info: "", backend: Gl } [wasm.js:344:9](http://127.0.0.1:1334/api/wasm.js)
WARN /home/duncan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bevy_pbr-0.12.0/src/ssao/mod.rs:87 ScreenSpaceAmbientOcclusionPlugin not loaded. GPU lacks support: TextureFormat::R16Float does not support TextureUsages::STORAGE_BINDING. [wasm.js:344:9](http://127.0.0.1:1334/api/wasm.js)
Uncaught (in promise) Error: Using exceptions for control flow, don't mind me. This isn't actually an error!
    __wbindgen_throw http://127.0.0.1:1334/api/wasm.js:1666
    __wbg_finalize_init http://127.0.0.1:1334/api/wasm.js:1743
    __wbg_init http://127.0.0.1:1334/api/wasm.js:1777
[wasm.js:1666:7](http://127.0.0.1:1334/api/wasm.js)
WebGL warning: compressedTexSubImage: Format does not allow partial sub-image updates. 8
ERROR  Shader translation error for stage ShaderStages(VERTEX): ERROR: 0:324: 'dFdx' : no matching overloaded function found
ERROR: 0:324: '=' : dimension mismatch
ERROR: 0:324: '=' : cannot convert from 'const mediump float' to 'highp 3-component vector of float'
ERROR: 0:325: 'dFdy' : no matching overloaded function found
ERROR: 0:325: '=' : dimension mismatch
ERROR: 0:325: '=' : cannot convert from 'const mediump float' to 'highp 3-component vector of float'
ERROR: 0:598: 'discard' : discard supported in fragment shaders only
ERROR: 0:620: 'unpackUnorm4x8' : no matching overloaded function found
ERROR: 0:620: '=' : dimension mismatch
ERROR: 0:620: 'assign' : cannot convert from 'const mediump float' to 'highp 4-component vector of float'
 log.target = "wgpu::backend::direct";
log.module_path = "wgpu::backend::direct";
log.file = "/home/duncan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-0.17.1/src/backend/direct.rs";
log.line = 1211; [wasm.js:344:9](http://127.0.0.1:1334/api/wasm.js)
ERROR  Please report it to https://github.com/gfx-rs/naga log.target = "wgpu::backend::direct";
log.module_path = "wgpu::backend::direct";
log.file = "/home/duncan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-0.17.1/src/backend/direct.rs";
log.line = 1212; [wasm.js:344:9](http://127.0.0.1:1334/api/wasm.js)
ERROR  Handling wgpu errors as fatal by default log.target = "wgpu::backend::direct";
log.module_path = "wgpu::backend::direct";
log.file = "/home/duncan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-0.17.1/src/backend/direct.rs";
log.line = 3055; [wasm.js:344:9](http://127.0.0.1:1334/api/wasm.js)
panicked at /home/duncan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-0.17.1/src/backend/direct.rs:3056:5:
wgpu error: Validation Error

Caused by:
    In Device::create_render_pipeline
      note: label = `opaque_mesh_pipeline`
    Internal error in ShaderStages(VERTEX) shader: ERROR: 0:324: 'dFdx' : no matching overloaded function found
ERROR: 0:324: '=' : dimension mismatch
ERROR: 0:324: '=' : cannot convert from 'const mediump float' to 'highp 3-component vector of float'
ERROR: 0:325: 'dFdy' : no matching overloaded function found
ERROR: 0:325: '=' : dimension mismatch
ERROR: 0:325: '=' : cannot convert from 'const mediump float' to 'highp 3-component vector of float'
ERROR: 0:598: 'discard' : discard supported in fragment shaders only
ERROR: 0:620: 'unpackUnorm4x8' : no matching overloaded function found
ERROR: 0:620: '=' : dimension mismatch
ERROR: 0:620: 'assign' : cannot convert from 'const mediump float' to 'highp 4-component vector of float'




Stack:

__wbg_get_imports/imports.wbg.__wbg_new_abda76e883ba8a5f@http://127.0.0.1:1334/api/wasm.js:369:13
__wbg_new_abda76e883ba8a5f externref shim@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[30106]:0xe9a695
console_error_panic_hook::hook::h3bd5cabdf97bcf51@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[8329]:0xb075ba
core::ops::function::Fn::call::h941e1848f1ab9e6d@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[42048]:0xed2a65
std::panicking::rust_panic_with_hook::h3aa054d35a0817d7@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[14898]:0xd382e3
std::panicking::begin_panic_handler::{{closure}}::h2f73e4cf6cd6319a@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[17044]:0xd9d8fe
std::sys_common::backtrace::__rust_end_short_backtrace::h98ac61a6abbff7e9@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[41592]:0xed1a5f
rust_begin_unwind@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[21175]:0xe24d51
core::panicking::panic_fmt::h3e1dd3d08288569e@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[24122]:0xe60b6f
core::ops::function::Fn::call::h9b50277317f455bb@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[14498]:0xd21305
wgpu::backend::direct::ErrorSinkRaw::handle_error::hbf2ff519dc8b30d6@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[11321]:0xc4569f
<wgpu::backend::direct::Context as wgpu::context::Context>::device_create_render_pipeline::h05fad58958bf0d80@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[716]:0x2e04bc
<T as wgpu::context::DynContext>::device_create_render_pipeline::h997cdd22f974389c@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[20936]:0xe1ecf9
wgpu::Device::create_render_pipeline::h2ea9abb2310067ba@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[16537]:0xd87d05
bevy_render::render_resource::pipeline_cache::PipelineCache::process_queue::h40dfb80d9f595a10@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[577]:0x245f51
bevy_ecs::system::system::System::run::h200faf57f01660d3@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[10276]:0xbe2f38
<bevy_ecs::schedule::executor::single_threaded::SingleThreadedExecutor as bevy_ecs::schedule::executor::SystemExecutor>::run::h782acc6a9978a8db@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[1298]:0x46cf64
bevy_ecs::schedule::schedule::Schedule::run::hc1573d6b31acf9e6@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[13227]:0xcd640c
bevy_ecs::world::World::schedule_scope::h56b6457ead9d16e1@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[2603]:0x694c41
bevy_app::app::App::update::hc45d58bb32d2a4ab@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[13482]:0xce70b4
winit::platform_impl::platform::event_loop::EventLoop<T>::spawn::{{closure}}::hc8f6dd95cd4964c4@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[442]:0x14e7bd
winit::platform_impl::platform::event_loop::runner::Shared<T>::handle_event::h0847cf5edfeae9e6@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[2820]:0x6da690
winit::platform_impl::platform::event_loop::runner::Shared<T>::run_until_cleared::hc955fde324a27044@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[4263]:0x859a1a
winit::platform_impl::platform::backend::timeout::AnimationFrameRequest::new::{{closure}}::h4db9e84169066ff2@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[23781]:0xe5b411
<dyn core::ops::function::FnMut<()>+Output = R as wasm_bindgen::closure::WasmClosure>::describe::invoke::h925a06056598a264@http://127.0.0.1:1334/api/wasm.wasm:wasm-function[27247]:0xe854f2
__wbg_adapter_39@http://127.0.0.1:1334/api/wasm.js:196:6
real@http://127.0.0.1:1334/api/wasm.js:177:32
```

**Comment by bonsairobo** [comment]
https://github.com/gfx-rs/wgpu/issues/4368

**Comment by bonsairobo** [comment]
Fixed by d43c203cb61f31056ab87bbe7b94ae49c5c32b4a
